storm-core/src/jvm/backtype/storm/blobstore/KeySequenceNumber.java

```diff
@@ -0,0 +1,227 @@
+ */
+public class KeySequenceNumber {
+    private static final Logger LOG = LoggerFactory.getLogger(Utils.class);
+    private final String BLOBSTORE_SUBTREE="/blobstore";
+    private final String BLOBSTORE_MAX_KEY_SEQUENCE_SUBTREE="/blobstoremaxkeysequencenumber";
+    private final String key;
+    private final NimbusInfo nimbusInfo;
+    private final int INT_CAPACITY = 4;
+
+    public KeySequenceNumber(String key, NimbusInfo nimbusInfo) {
+        this.key = key;
+        this.nimbusInfo = nimbusInfo;
+    }
+
+    public int getKeySequenceNumber(Map conf) {
+        TreeSet<Integer> sequenceNumbers = new TreeSet<Integer>();
+        CuratorFramework zkClient = BlobStoreUtils.createZKClient(conf);
+        try {
+            // Key has not been created yet and it is the first time it is being created
+            if(zkClient.checkExists().forPath(BLOBSTORE_SUBTREE + "/" + key) == null) {
+                zkClient.create().creatingParentsIfNeeded().withMode(CreateMode.PERSISTENT)
+                        .withACL(ZooDefs.Ids.OPEN_ACL_UNSAFE).forPath(BLOBSTORE_MAX_KEY_SEQUENCE_SUBTREE + "/" + key);
+                zkClient.setData().forPath(BLOBSTORE_MAX_KEY_SEQUENCE_SUBTREE + "/" + key,
+                        ByteBuffer.allocate(INT_CAPACITY).putInt(1).array());
+                return 1;
+            }
+
+            // When all nimbodes go down and one or few of them come up
+            // Unfortunately there might not be an exact way to know which one contains the most updated blob,
+            // if all go down which is unlikely. Hence there might be a need to update the blob if all go down.
+            List<String> stateInfoList = zkClient.getChildren().forPath(BLOBSTORE_SUBTREE + "/" + key);
+            LOG.debug("stateInfoList-size {} stateInfoList-data {}", stateInfoList.size(), stateInfoList);
+            if(stateInfoList.isEmpty()) {
+                return getMaxSequenceNumber(zkClient);
+            }
+
+            LOG.debug("stateInfoSize {}", stateInfoList.size());
+            // In all other cases check for the latest update sequence of the blob on the nimbus
+            // and assign the appropriate number. Check if all are have same sequence number,
+            // if not assign the highest sequence number.
+            for (String stateInfo:stateInfoList) {
+                sequenceNumbers.add(Integer.parseInt(BlobStoreUtils.normalizeVersionInfo(stateInfo)[1]));
+            }
+
+            // Update scenario 2 and 3 explain the code logic written here
+            // especially when nimbus crashes and comes up after and before update
+            // respectively.
+            int currentSeqNumber = getMaxSequenceNumber(zkClient);
+            if (!checkIfStateContainsCurrentNimbusHost(stateInfoList, nimbusInfo) && !nimbusInfo.isLeader()) {
+                if (sequenceNumbers.last() < currentSeqNumber) {
+                    return currentSeqNumber;
+                } else {
+                    return 0;
```

Comment:
Could return `INITIAL_SEQUENCE_NUMBER - 1`, so it is clearly guaranteed that the caller will not think its copy of this key's blob is current.